src/System.IO.Pipelines/PipeCompletion.cs

```diff
@@ -1,58 +1,137 @@
 ﻿// Copyright (c) Microsoft. All rights reserved.
 // Licensed under the MIT license. See LICENSE file in the project root for full license information.
 
-using System.Threading;
+using System.Diagnostics;
+using System.Runtime.CompilerServices;
 
 namespace System.IO.Pipelines
 {
     internal struct PipeCompletion
     {
+        private const int InitialCallbacksSize = 1;
         private static readonly Exception _completedNoException = new Exception();
 
-        private Exception _exception;
 #if COMPLETION_LOCATION_TRACKING
         private string _completionLocation;
 #endif
+        public Exception Exception { get; set; }
 
-        public bool IsCompleted => _exception != null;
+        private PipeCompletionCallback[] _callbacks;
+        private int _callbackCount;
 
-        public void TryComplete(Exception exception = null)
+        public string Location
+        {
+            get
+            {
+#if COMPLETION_LOCATION_TRACKING
+                return _completionLocation;
+#else
+                return null;
+#endif
+            }
+        }
+
+        public bool IsCompleted => Exception != null;
+
+        public bool TryComplete(Exception exception = null)
         {
 #if COMPLETION_LOCATION_TRACKING
             _completionLocation = Environment.StackTrace;
 #endif
-            // Set the exception object to the exception passed in or a sentinel value
-            Interlocked.CompareExchange(ref _exception, exception ?? _completedNoException, null);
+            if (Exception == null)
+            {
+                // Set the exception object to the exception passed in or a sentinel value
+                Exception = exception ?? _completedNoException;
+            }
+            return _callbackCount > 0;
         }
 
-        public bool IsCompletedOrThrow()
+        public bool AddCallback(Action<Exception, object> callback, object state)
         {
-            if (_exception != null)
+            if (IsCompleted)
             {
-                if (_exception != _completedNoException)
-                {
-                    ThrowFailed();
-                }
                 return true;
             }
+
+            if (_callbacks == null)
+            {
+                _callbacks = new PipeCompletionCallback[InitialCallbacksSize];
+            }
+
+            var newIndex = _callbackCount;
+            _callbackCount++;
+
+            if (_callbackCount == _callbacks.Length)
```

Comment:
This is particularly funny now that InitialCallbacksSize is 1. The first callback to be added causes a 1 element array to be allocated only to be immediately discarded to be replaced with a 2 element array. :rofl: